init/devices.cpp

```diff
@@ -399,44 +415,110 @@ std::vector<std::string> DeviceHandler::GetBlockDeviceSymlinks(const Uevent& uev
     return links;
 }
 
-void DeviceHandler::HandleDevice(const std::string& action, const std::string& devpath,
+void DeviceHandler::HandleDevice(const std::string& action, const std::string& dev_path,
                                  const std::string& upath, bool block,
                                  int major, int minor, const std::vector<std::string>& links) const {
-    std::vector<std::string> all_links(links);
-    std::string alias_link;
+    std::string state_link_path, alias_link_path;
+    std::stringstream slfmt;
 
-    // Add the alias link to the device to the list of links to
-    // create.
-    if (GetDeviceAlias(upath, major, minor, alias_link))
-        all_links.push_back(alias_link);
+    // Create a symlink in /dev/aliases that will point to the
+    // alias.
+    slfmt << "/dev/aliases/" << major << "_" << minor;
+    state_link_path = slfmt.str();
 
     if (action == "add") {
-        MakeDevice(devpath, block, major, minor, all_links);
-        for (const auto& link : all_links) {
-            if (!mkdir_recursive(Dirname(link), 0755)) {
-                PLOG(ERROR) << "Failed to create directory " << Dirname(link);
+        std::string alias_path;
+        std::vector<std::string> all_links(links);
+
+        // Add the alias link to the device to the list of links to
+        // create.
+        if (GetDeviceAlias(upath, major, minor, alias_link_path))
+            all_links.push_back(alias_link_path);
+
+        MakeDevice(dev_path, block, major, minor, all_links);
+
+        for (const auto& link_path : all_links) {
+            std::string link_target;
+
+            if (!mkdir_recursive(Dirname(link_path), 0755))
+                PLOG(ERROR) << "Failed to create directory " << Dirname(link_path);
+
+            // Only create a link in /dev/aliases for the device
+            // aliases we created from an 'alias' line in the
+            // configuration. This is a way to save what aliases we've
+            // created in /dev for a specific device and allow use to
+            // find the exact aliases we created without keeping state
+            // within ueventd.
+
+            // alias_link is the link (eg: /dev/modem -> /dev/ttyACM0)
+            // that we want to be creating.
+            if (!alias_link_path.empty()
+                && link_path == alias_link_path
+                && symlink(alias_link_path.c_str(), state_link_path.c_str())) {
```

Comment:
I must be missing something but when is `alias_link_path` created as we just create `state_link_path` here in which case the `else` part would be skipped.

Reply:
I was confused for a second at my code but then remembered the catch. The first branch within the if will be executed...

IF there is an alias to create (!alias_link.empty())
IF the iteration through the links has reached the alias to create (link_path == alias_link_path)
IF the call to symlink **FAILS** (ie: symlink() returns 0 on success)

Reply:
Right. Short circuit evaluation is where we decide if we're creating the symlink or not. And else on success. Makes sense.